lib/Corres_UL.thy

```diff
@@ -905,12 +905,29 @@ qed
 lemma corres_inst: "corres_underlying sr nf nf' r P P' f g \<Longrightarrow> corres_underlying sr nf nf' r P P' f g" .
 
 lemma corres_assert_opt_assume:
-  assumes "\<And>x. P' = Some x \<Longrightarrow> corres_underlying sr nf nf' r P Q f (g x)"
-  assumes "\<And>s. Q s \<Longrightarrow> P' \<noteq> None"
-  shows "corres_underlying sr nf nf' r P Q f (assert_opt P' >>= g)" using assms
+  assumes "\<And>x. opt = Some x \<Longrightarrow> corres_underlying sr nf nf' r P Q f (g x)"
+  assumes "\<And>s. Q s \<Longrightarrow> opt \<noteq> None"
+  shows "corres_underlying sr nf nf' r P Q f (assert_opt opt >>= g)" using assms
   by (auto simp: bind_def assert_opt_def assert_def fail_def return_def
                  corres_underlying_def split: option.splits)
 
+lemma corres_assert_opt_assume':
+  "(\<And>x. opt = Some x \<Longrightarrow> corres_underlying sr nf nf' r P Q f (g x))
+   \<Longrightarrow> corres_underlying sr nf nf' r P (Q and K (opt \<noteq> None)) f (assert_opt opt >>= g)"
+  by (fastforce intro: corres_gen_asm2 stronger_corres_guard_imp corres_assert_opt_assume)
+
+lemma corres_assert_opt_assume_lhs:
+  assumes "\<And>x. opt = Some x \<Longrightarrow> corres_underlying sr nf nf' r P Q (f x) g"
+  assumes "\<And>s. P s \<Longrightarrow> opt \<noteq> None"
+  shows "corres_underlying sr nf nf' r P Q (assert_opt opt >>= f) g" using assms
+  by (auto simp: bind_def assert_opt_def assert_def fail_def return_def
+                 corres_underlying_def split: option.splits)
```

Comment:
style: `split:` should go on separate line

Reply:
actually, this style is kinda wonky, the using assms at the end of the line is not great, but also pointless
Could you consider rewriting this to `[| ...; ...|] ==> ...` form?

Reply:
`fastforce` could actually take care of all of these with fewer rules for `simp:` and without the `split:`, which is nice.